File: AndroidStudioProjects/HealLotsMobile/app/src/main/java/com/heallots/mobile/ui/adapters/ReviewAdapter.java
package com.heallots.mobile.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.heallots.mobile.R;
import com.heallots.mobile.models.Review;
import com.heallots.mobile.utils.Constants;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class ReviewAdapter extends RecyclerView.Adapter<ReviewAdapter.ReviewViewHolder> {
    private final Context context;
    private List<Review> reviews = new ArrayList<>();

    public ReviewAdapter(Context context, List<Review> reviews) {
        this.context = context;
        this.reviews = reviews != null ? reviews : new ArrayList<>();
    }

    public void updateReviews(List<Review> updatedReviews) {
        this.reviews = updatedReviews != null ? updatedReviews : new ArrayList<>();
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ReviewViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_dashboard_review, parent, false);
        return new ReviewViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReviewViewHolder holder, int position) {
        holder.bind(reviews.get(position));
    }

    @Override
    public int getItemCount() {
        return reviews.size();
    }

    class ReviewViewHolder extends RecyclerView.ViewHolder {
        private final ImageView avatar;
        private final TextView avatarText;
        private final TextView patientName;
        private final TextView rating;
        private final TextView meta;
        private final TextView body;

        ReviewViewHolder(@NonNull View itemView) {
            super(itemView);
            avatar = itemView.findViewById(R.id.reviewAvatar);
            avatarText = itemView.findViewById(R.id.reviewAvatarText);
            patientName = itemView.findViewById(R.id.reviewPatientName);
            rating = itemView.findViewById(R.id.reviewRating);
            meta = itemView.findViewById(R.id.reviewMeta);
            body = itemView.findViewById(R.id.reviewBody);
        }

        void bind(Review review) {
            String name = safeText(review.getPatientName(), "Guest");
            patientName.setText(name);
            avatarText.setText(name.substring(0, 1).toUpperCase());
            rating.setText(buildStars(review.getRating()));
            meta.setText(safeText(review.getServiceName(), "Hilot Session") + " with " + safeText(review.getSpecialistName(), "Specialist"));
            body.setText(safeText(review.getReviewText(), "Shared a positive experience."));

            String profilePicture = review.getPatientProfilePictureUrl();
            if (profilePicture != null && !profilePicture.trim().isEmpty()) {
                avatar.setVisibility(View.VISIBLE);
                avatarText.setVisibility(View.VISIBLE);
                String imageUrl = profilePicture.startsWith("http")
                        ? profilePicture
                        : Constants.BASE_URL + "/api/user/profile-picture/" + profilePicture;
                Picasso.get().load(imageUrl).fit().centerCrop().into(avatar, new com.squareup.picasso.Callback() {
                    @Override
                    public void onSuccess() {
                        avatarText.setVisibility(View.GONE);
                    }

                    @Override
                    public void onError(Exception e) {
                        avatar.setImageDrawable(null);
                        avatarText.setVisibility(View.VISIBLE);
                    }
                });
            } else {
                avatar.setImageDrawable(null);
                avatarText.setVisibility(View.VISIBLE);
            }
        }

        private String buildStars(int stars) {
            int safeStars = Math.max(1, Math.min(stars, 5));
            StringBuilder builder = new StringBuilder();
            for (int i = 0; i < safeStars; i++) {
                builder.append("★");
            }
            return builder.toString();
        }

        private String safeText(String value, String fallback) {
            return value == null || value.trim().isEmpty() ? fallback : value;
        }
    }
}

File: AndroidStudioProjects/HealLotsMobile/app/src/main/java/com/heallots/mobile/ui/adapters/ReviewAdapter.kt
package com.heallots.mobile.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.heallots.mobile.R
import com.heallots.mobile.models.Review
import com.heallots.mobile.utils.Constants
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import java.util.Locale

class ReviewAdapter(
    private val context: Context,
    reviews: List<Review>?
) : RecyclerView.Adapter<ReviewAdapter.ReviewViewHolder>() {
    private var reviews: List<Review> = reviews ?: ArrayList()

    fun updateReviews(updatedReviews: List<Review>?) {
        reviews = updatedReviews ?: ArrayList()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReviewViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_dashboard_review, parent, false)
        return ReviewViewHolder(view)
    }

    override fun onBindViewHolder(holder: ReviewViewHolder, position: Int) {
        holder.bind(reviews[position])
    }

    override fun getItemCount(): Int = reviews.size

    inner class ReviewViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val avatar: ImageView = itemView.findViewById(R.id.reviewAvatar)
        private val avatarText: TextView = itemView.findViewById(R.id.reviewAvatarText)
        private val patientName: TextView = itemView.findViewById(R.id.reviewPatientName)
        private val rating: TextView = itemView.findViewById(R.id.reviewRating)
        private val meta: TextView = itemView.findViewById(R.id.reviewMeta)
        private val body: TextView = itemView.findViewById(R.id.reviewBody)

        fun bind(review: Review) {
            val name = safeText(review.patientName, "Guest")
            patientName.text = name
            avatarText.text = name.substring(0, 1).uppercase(Locale.getDefault())
            rating.text = buildStars(review.rating)
            meta.text = "${safeText(review.serviceName, "Hilot Session")} with ${safeText(review.specialistName, "Specialist")}"
            body.text = safeText(review.reviewText, "Shared a positive experience.")

            val profilePicture = review.patientProfilePictureUrl
            if (!profilePicture.isNullOrBlank()) {
                avatar.visibility = View.VISIBLE
                avatarText.visibility = View.VISIBLE
                val imageUrl = if (profilePicture.startsWith("http")) {
                    profilePicture
                } else {
                    "${Constants.BASE_URL}/api/user/profile-picture/$profilePicture"
                }
                Picasso.get().load(imageUrl).fit().centerCrop().into(avatar, object : Callback {
                    override fun onSuccess() {
                        avatarText.visibility = View.GONE
                    }

                    override fun onError(e: Exception?) {
                        avatar.setImageDrawable(null)
                        avatarText.visibility = View.VISIBLE
                    }
                })
            } else {
                avatar.setImageDrawable(null)
                avatarText.visibility = View.VISIBLE
            }
        }

        private fun buildStars(stars: Int): String {
            val safeStars = stars.coerceIn(1, 5)
            return buildString {
                repeat(safeStars) {
                    append('\u2605')
                }
            }
        }

        private fun safeText(value: String?, fallback: String): String {
            return if (value.isNullOrBlank()) fallback else value
        }
    }
}
